lib/preview-web/src/PreviewWeb.tsx

```diff
@@ -0,0 +1,559 @@
+    this.storyStore.globals.update(globals);
+
+    this.channel.emit(Events.GLOBALS_UPDATED, {
+      globals: this.storyStore.globals.get(),
+      initialGlobals: this.storyStore.globals.initialGlobals,
+    });
+  }
+
+  onUpdateArgs({ storyId, updatedArgs }: { storyId: StoryId; updatedArgs: Args }) {
+    this.storyStore.args.update(storyId, updatedArgs);
+    this.channel.emit(Events.STORY_ARGS_UPDATED, {
+      storyId,
+      args: this.storyStore.args.get(storyId),
+    });
+  }
+
+  async onResetArgs({ storyId, argNames }: { storyId: string; argNames?: string[] }) {
+    // NOTE: we have to be careful here and avoid await-ing when updating the current story's args.
+    // That's because below in `renderStoryToElement` we have also bound to this event and will
+    // render the story in the same tick.
+    // However, we can do that safely as the current story is available in `this.previousStory`
+    const { initialArgs } =
+      storyId === this.previousStory.id
+        ? this.previousStory
+        : await this.storyStore.loadStory({ storyId });
+
+    const argNamesToReset = argNames || Object.keys(this.storyStore.args.get(storyId));
+    const updatedArgs = argNamesToReset.reduce((acc, argName) => {
+      acc[argName] = initialArgs[argName];
+      return acc;
+    }, {} as Partial<Args>);
+
+    this.onUpdateArgs({ storyId, updatedArgs });
+  }
+
+  // This happens when a glob gets HMR-ed
+  async onImportFnChanged({ importFn }: { importFn: ModuleImportFn }) {
+    await this.storyStore.onImportFnChanged({ importFn });
+
+    if (this.urlStore.selection) {
+      this.renderSelection();
+    } else {
+      this.selectSpecifiedStory();
+    }
+
+    if (!FEATURES?.storyStoreV7) {
+      this.channel.emit(Events.SET_STORIES, await this.storyStore.getSetStoriesPayload());
+    }
+  }
+
+  // This happens when a config file gets reloade
+  onGetProjectAnnotationsChanged({
+    getProjectAnnotations,
+  }: {
+    getProjectAnnotations: () => ProjectAnnotations<TFramework>;
+  }) {
+    const projectAnnotations = this.getProjectAnnotationsOrRenderError(getProjectAnnotations);
+    if (!projectAnnotations) {
+      return;
+    }
+
+    this.storyStore.updateProjectAnnotations(projectAnnotations);
+    this.renderSelection();
+  }
+
+  // We can either have:
+  // - a story selected in "story" viewMode,
+  //     in which case we render it to the root element, OR
+  // - a story selected in "docs" viewMode,
+  //     in which case we render the docsPage for that story
+  async renderSelection({ persistedArgs }: { persistedArgs?: Args } = {}) {
+    if (!this.urlStore.selection) {
+      throw new Error('Cannot render story as no selection was made');
+    }
+
+    const { selection } = this.urlStore;
+
+    let story;
+    try {
+      story = await this.storyStore.loadStory({ storyId: selection.storyId });
+    } catch (err) {
+      logger.warn(err);
+      this.renderMissingStory(selection.storyId);
```

Comment:
I guess this is more of a feature request than anything, I'm just wondering if it's possible:

Take the following example: a `Button` component with a `Default` story, which is loaded in a path like `?path=/story/button--default`.

Then, the user renames that story to something else, like `Primary`.

As a result, Storybook reloads, but the path is still `button--default` and not `button-primary`, so the "no preview" error will be shown. This behaviour can be a bit frustrating, and happens in current Storybook.

Would it make sense to:
**1 - Detect such scenario and try to select the path with the new id of the story.** 
The heuristic for this could be a bit complicated and likely not always precise, because people could be deleting a story instead of renaming, altering multiple story names etc.

**2 - Update the text of the "No preview" component to mention renaming or deleting a story as a possible cause**
Simple change which I think adds value

**3 - Keep things as is**

Reply:
FWIW, the current mechanism for preserving the URL is:

```js
export const Default = {
  name: 'Primary',
  args: ...
}
```

Reply:
@yannbf I think we could probably detect when the "story missing" happens on HMR from a rendered story, so even though we probably can't tell what the new story is, we could make a separate message.

Reply:
What would be the difference between "story missing" and "no preview"? Is it that the "story missing" stories have a known component and the "no preview" stories don't?

Reply:
We currently show this:

<img width="752" alt="image" src="https://user-images.githubusercontent.com/132554/132614193-ed8eca4f-eeae-400d-951e-d8739388273e.png">

Both in the case that (a) you pass a "selector" that matches nothing (e.g. `*` when there are no stories at all)
(b) you pass a story id for a story that doesn't exist.

I would suggest in the case (b) that happens on HMR where `previousStory` exists and matches the storyId, we tweak the message to indicate the story's probably gone due to HMR